prov/gni/src/gnix_nic.c

```diff
@@ -333,30 +338,38 @@ static int __process_rx_cqe(struct gnix_nic *nic, gni_cq_entry_t cqe)
 	struct gnix_vc *vc;
 
 	vc_id =  GNI_CQ_GET_INST_ID(cqe);
-	vc = __gnix_nic_elem_by_rem_id(nic, vc_id);
 
-#if 1 /* Process RX inline with arrival of an RX CQE. */
-	if (unlikely(vc->conn_state != GNIX_VC_CONNECTED)) {
-		GNIX_DEBUG(FI_LOG_EP_DATA,
-			  "Scheduling VC for RX processing (%p)\n",
-			  vc);
-		ret = _gnix_vc_rx_schedule(vc);
-		assert(ret == FI_SUCCESS);
-	} else {
-		GNIX_DEBUG(FI_LOG_EP_DATA,
-			  "Processing VC RX (%p)\n",
-			  vc);
-		ret = _gnix_vc_dequeue_smsg(vc);
-		if (ret != FI_SUCCESS) {
-			GNIX_WARN(FI_LOG_EP_DATA,
+	/*
+	 * its possible this vc has been destroyed, so may get NULL
+	 * back.
+	 */
+
+	vc = __gnix_nic_elem_by_rem_id(nic, vc_id);
+	if (vc != NULL) {
```

Comment:
Thanks for explanation.  So my last request before merging is that you add a comment at the declaration of the freelist in gnix_nic.  It should say what you explained above, so that somewhere down the line, someone doesn't come along and replace the freelist with a different type of allocator that might not have the same properties (the most important being that the allocated elements of the freelist are not ever returned to the system).